contracts/LockedNORIV2.sol

```diff
@@ -0,0 +1,791 @@
+ *
+ * ##### Implements
+ *
+ * - [IERC777RecipientUpgradeable](https://docs.openzeppelin.com/contracts/4.x/api/token/erc777#IERC777Recipient)
+ * - [IERC777Upgradeable](https://docs.openzeppelin.com/contracts/4.x/api/token/erc777#IERC777)
+ * - [IERC20Upgradeable](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20)
+ * - [IAccessControlEnumerable](https://docs.openzeppelin.com/contracts/4.x/api/access#AccessControlEnumerable)
+ * - [IERC165Upgradeable](https://docs.openzeppelin.com/contracts/4.x/api/utils#IERC165)
+ *
+ * ##### Uses
+ *
+ * - [ScheduleUtils](./ScheduleUtils.md) for Schedule
+ * - [MathUpgradeable](https://docs.openzeppelin.com/contracts/4.x/api/utils#Math)
+ *
+ */
+contract LockedNORIV2 is
+  IERC777RecipientUpgradeable,
+  ERC777PresetPausablePermissioned
+{
+  using ScheduleUtils for Schedule;
+
+  struct TokenGrant {
+    Schedule vestingSchedule;
+    Schedule lockupSchedule;
+    uint256 grantAmount;
+    uint256 claimedAmount;
+    uint256 originalAmount;
+    bool exists;
+    uint256 lastRevocationTime;
+    uint256 lastQuantityRevoked;
+  }
+
+  struct TokenGrantDetail {
+    uint256 grantAmount;
+    address recipient;
+    uint256 startTime;
+    uint256 vestEndTime;
+    uint256 unlockEndTime;
+    uint256 cliff1Time;
+    uint256 cliff2Time;
+    uint256 vestCliff1Amount;
+    uint256 vestCliff2Amount;
+    uint256 unlockCliff1Amount;
+    uint256 unlockCliff2Amount;
+    uint256 claimedAmount;
+    uint256 originalAmount;
+    uint256 lastRevocationTime;
+    uint256 lastQuantityRevoked;
+    bool exists;
+  }
+
+  struct CreateTokenGrantParams {
+    address recipient;
+    uint256 startTime;
+    uint256 vestEndTime;
+    uint256 unlockEndTime;
+    uint256 cliff1Time;
+    uint256 cliff2Time;
+    uint256 vestCliff1Amount;
+    uint256 vestCliff2Amount;
+    uint256 unlockCliff1Amount;
+    uint256 unlockCliff2Amount;
+  }
+
+  struct DepositForParams {
+    address recipient;
+    uint256 startTime;
+  }
+
+  /**
+   * @notice Role conferring creation and revocation of token grants.
+   */
+  bytes32 public constant TOKEN_GRANTER_ROLE = keccak256("TOKEN_GRANTER_ROLE");
+
+  /**
+   * @notice Used to register the ERC777TokensRecipient recipient interface in the
+   * ERC-1820 registry
+   *
+   * @dev Registering that LockedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  bytes32 public constant ERC777_TOKENS_RECIPIENT_HASH =
+    keccak256("ERC777TokensRecipient");
+
+  /**
+   * @notice A mapping from grantee to grant
+   */
+  mapping(address => TokenGrant) private _grants;
+
+  /**
+   * @notice The BridgedPolygonNORI contract that this contract wraps tokens for
+   */
+  IERC20Upgradeable private _bridgedPolygonNori;
```

Comment:
Sure, could do.  I figured it might be more self-explanatory / self-contained this way since we only interact with BridgedPolygonNORI as an ERC20 but I'm not strongly attached to that idea.

I'm aware that once I start fixing the vesting task we may want to add bulk grant creation and deposit functions in here which may change that story.